### PR TITLE
v0.2 - refactoring

### PR DESCRIPTION
**Pull request by Crimscent**

### CODE REVIEW

File: config-core/src/main/java/com/revolut/config/Config.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    public String get(String key, String separator) {
+        return get(configKey(key, separator));
+    }
+
+    public String get(ConfigKey key) {
+        for (int i = configProviders.size() - 1; i >= 0; i--) {
+            final ConfigProvider configProvider = configProviders.get(i);
+            if (configProvider.has(key)) {
+                return configProvider.get(key);
```
Comment (spolnik) [inline config-core/src/main/java/com/revolut/config/Config.java]: that can be done with optional being returned from configProvider, and no need for two methdos then - you can use map and streams

File: config-core/src/main/java/com/revolut/config/Config.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            }
+        }
+        return null;
```
Comment (spolnik) [inline config-core/src/main/java/com/revolut/config/Config.java]: absent() vs null please

File: config-core/src/main/java/com/revolut/config/Config.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public Map<ConfigKey, String> getAll() {
+        final Map<ConfigKey, String> map = new TreeMap<>();
+        for (final ConfigProvider configProvider : configProviders) {
+            configProvider.getAll().forEach(map::put);
```
Comment (spolnik) [inline config-core/src/main/java/com/revolut/config/Config.java]: how about `TreeMap#putAll` ?

File: config-core/src/main/java/com/revolut/config/Config.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        return map;
+    }
+
+    public <T> T bindUsing(ConfigMapper configMapper, Class<T> tClass) {
+        return configMapper.map(this, tClass);
```
Comment (spolnik) [inline config-core/src/main/java/com/revolut/config/Config.java]: please rename `tClass` to `type` or `klass`

File: config-core/src/main/java/com/revolut/config/mappers/ConfigMapper.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.revolut.config.mappers;
+
+import com.revolut.config.Config;
+
+public interface ConfigMapper {
+
+    <T> T map(Config config, Class<T> tClass);
```
Comment (spolnik) [inline config-core/src/main/java/com/revolut/config/mappers/ConfigMapper.java]: please rename `tClass` to `klass` or `type`

File: config-core/src/main/java/com/revolut/config/providers/ConfigProvider.java [hunk 6]
```diff
@@ -2,15 +2,15 @@
...
 
-import static com.revolut.config.helpers.Validation.validateNotNull;
+import static com.revolut.config.helpers.Guards.requireNotNull;
 import static java.util.Collections.unmodifiableMap;
 
-public abstract class ConfigurationProvider {
+public abstract class ConfigProvider {
 
-    private Map<ConfigKey, String> map = new HashMap<>();
+    private Map<ConfigKey, String> map = new TreeMap<>();
```
Comment (spolnik) [inline config-core/src/main/java/com/revolut/config/providers/ConfigProvider.java]: why do we need to keep sorted map here 🤔 do we somehow use sub-maps here?

Comment (Crimscent) [reply config-core/src/main/java/com/revolut/config/providers/ConfigProvider.java]: No, for consistency with Config.

**Pull request merged**